src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -154,9 +145,76 @@ internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<
                 options.SuppressEnvironmentConfiguration = true;
             });
 
+        _genericWebHostServiceDescriptor = InitializeHosting(bootstrapHostBuilder);
+    }
+
+    internal WebApplicationBuilder(WebApplicationOptions options, bool slim, bool empty, Action<IHostBuilder>? configureDefaults = null)
+    {
+        Debug.Assert(!slim, "should only be called with slim: false");
+        Debug.Assert(empty, "should only be called with empty: true");
+
+        var configuration = new ConfigurationManager();
+
+        // empty builder should still default the ContentRoot as usual. This is the expected behavior for all WebApplicationBuilders.
+        SetDefaultContentRoot(options, configuration);
+
+        _hostApplicationBuilder = Microsoft.Extensions.Hosting.Host.CreateEmptyApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            Args = options.Args,
+            ApplicationName = options.ApplicationName,
+            EnvironmentName = options.EnvironmentName,
+            ContentRootPath = options.ContentRootPath,
+            Configuration = configuration,
+        });
+
+        // Set WebRootPath if necessary
+        if (options.WebRootPath is not null)
+        {
+            Configuration.AddInMemoryCollection(new[]
+            {
+                new KeyValuePair<string, string?>(WebHostDefaults.WebRootKey, options.WebRootPath),
+            });
+        }
+
+        // Run methods to configure web host defaults early to populate services
+        var bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // This is for testing purposes
+        configureDefaults?.Invoke(bootstrapHostBuilder);
+
+        bootstrapHostBuilder.ConfigureSlimWebHost(
```

Comment:
I forgot about this public API. It is a little weird that the "empty" builder configures the "slim" web host.

As a follow up, would it make more sense to rename this to `ConfigureEmptyWebHost` and then add a `ConfigureSlimWebHost` that also calls `ConfigureWebDefaultsSlim` and adds Kestrel and the core routing services?

Reply:
I've opened https://github.com/dotnet/aspnetcore/issues/49375 to track this.

Reply:
Note that these are the things added in ConfigureSlimWebHost:

https://github.com/dotnet/aspnetcore/blob/224026829144defe56e878f4011b7a54ca10353c/src/Hosting/Hosting/src/GenericHost/SlimWebHostBuilder.cs#L20-L60

It doesn't do anything with Kestrel or routing. That is higher up in the stack.

Looking through these services, the only ones I could imagine removing in an "EmptyWebHost" are all related to observability:

* DiagnosticListener
* ActivitySource
* DistributedContextPropagator
* AddMetrics/HostingMetrics

Reply:
> It doesn't do anything with Kestrel or routing. That is higher up in the stack.

I realize it's that way now. I was suggesting moving `ConfigureWebDefaultsSlim` into `ConfigureSlimWebHost` as part of the change.

Reply:
It can't be because they live in different assemblies - thus why we needed to make that API public.

For example `ConfigureWebDefaultsSlim` uses stuff in `Microsoft.AspNetCore.dll`, like `HostFilteringStartupFilter` and in references like Routing, which `Microsoft.AspNetCore.Hosting.dll` doesn't have.